tsdb/querier_test.go

```diff
@@ -2194,6 +2195,76 @@ func TestPostingsForMatchers(t *testing.T) {
 	}
 }
 
+// TestQuerierIndexQueriesRace tests the index queries with racing appends.
+func TestQuerierIndexQueriesRace(t *testing.T) {
+	const testRepeats = 1000
+
+	testCases := []struct {
+		matchers []*labels.Matcher
+	}{
+		{
+			matchers: []*labels.Matcher{
+				// This matcher should involve the AllPostings posting list in calculating the posting lists.
+				labels.MustNewMatcher(labels.MatchNotEqual, labels.MetricName, "metric"),
+			},
+		},
+		{
+			matchers: []*labels.Matcher{
+				// The first matcher should be effectively the same as AllPostings, because all series have m=0
+				// If it is evaluated first, then __name__=metric will contain more series than m=0.
+				labels.MustNewMatcher(labels.MatchNotEqual, "m", "0"),
+				labels.MustNewMatcher(labels.MatchEqual, labels.MetricName, "metric"),
+			},
+		},
+	}
+
+	for _, c := range testCases {
+		c := c
+		t.Run(fmt.Sprintf("%v", c.matchers), func(t *testing.T) {
+			db := openTestDB(t, DefaultOptions(), nil)
+			h := db.Head()
+			t.Cleanup(func() {
+				require.NoError(t, db.Close())
+			})
+			ctx, cancel := context.WithCancel(context.Background())
+			wg := &sync.WaitGroup{}
+			wg.Add(1)
+			go appendSeries(t, ctx, wg, h)
+			t.Cleanup(wg.Wait)
+			t.Cleanup(cancel)
+
+			for i := 0; i < testRepeats; i++ {
+				q, err := db.Querier(ctx, math.MinInt64, math.MaxInt64)
+				require.NoError(t, err)
+
+				values, _, err := q.LabelValues("n", c.matchers...)
+				require.NoError(t, err)
+				require.Emptyf(t, values, `label values for label "n" should be empty`)
+			}
+		})
+	}
+}
+
+func appendSeries(t *testing.T, ctx context.Context, wg *sync.WaitGroup, h *Head) {
+	defer wg.Done()
+
+	for i := 0; ; i++ {
+		if ctx.Err() != nil {
+			return
+		}
+		app := h.Appender(context.Background())
+		_, err := app.Append(0, labels.FromStrings(labels.MetricName, "metric", "n", strconv.Itoa(i), "m", "0"), 0, 0)
+		require.NoError(t, err)
+		err = app.Commit()
+		require.NoError(t, err)
+
+		if i%10 == 0 {
+			// Throttle down the appends to keep the test somewhat nimble.
+			time.Sleep(10 * time.Millisecond)
+		}
```

Comment:
I wonder if just sleeping 1ms between each series would have a better effect. Because what we really want is to add a series at the same time when we're calculating postings. Adding 1 or 10 doesn't make the test different, but sleeping between each append makes it more prone to switch to scheduling the other goroutine.

Reply:
good call, with this change (sleep 1ms, every append) I'm able to reproduce the original bug slightly more often than with the previous setup (sleep 10ms, every 10 appends). Made the change in 5457f68e5